### PR TITLE
4.2 ml< >server utilities

### PR DESCRIPTION
**Pull request by nsokolnikov**
Classification blackbox completed
First model works and outputs things with properly formed requests

Support for model storing, as well as dynamic window sizing in the first model.

The window_size is the amount of data points used for classification. Large values would give kind of meaningless results as they are likely to encompass several running devices, and small values may not be as accurate. 

Model is a RandomForestModel from scikit-learn. The way to construct it is 
```model = RandomForestModel(window_size plus any arguments in baseClassifier class)```

Train it with
```model.train(your_data) ```
The format of the data is shown in the testdata variable in the source code for RandomForestClassifier. 

Classify with 
```model.classify(time) ```
It will automatically get the latest window_size rows since that timestamp and classify based on that. 

To do it as a request to the server:
```db.sead.systems/<device_id>/start_time=<start_time>&device=<Panel selection>&classify=1 ```

### CODE REVIEW

**Comment by aadabi** [comment]
All checks have failed?

**Comment by ianlofs** [comment]
@aadabi we probably need to add the sklearn dependencies to the Travis file

**Comment by iangudger** [comment]
Please fetch and merge the latest CI changes.

**Comment by iangudger** [comment]
The problem is that query_db.py imports a bunch of  DB.* packages and there isn't a package named DB (no file named DB.py in the directory).

**Comment by ianlofs** [comment]
All the imports are working now. Two tests are now failing test_classify and test_classify_missing_start_end_time. This make sense though because classification is the major change of this pull request. I will try to get the tests to pass.

**Comment by ianlofs** [comment]
Not really sure how to fix the last failing test, @iangudger do you have any ideas?

File: .travis.yml [hunk 1]
```diff
@@ -12,14 +12,17 @@ env:
     - BACKEND="$HOME/gopath/src/github.com/seadsystem/Backend"
     - API="$BACKEND/DB/api"
     - LZONE="$BACKEND/DB/landingzone"
+    - PYTHONPATH="$BACKEND"
```
Comment (iangudger) [inline .travis.yml]: Is this used?

File: DB/.gitignore [hunk 2]
```diff
@@ -1,8 +1,10 @@
 *.sw[po]
 *~
-api/__pycache__
+__pycache__
 tags
 .idea
 .DS_Store
 .vagrant
+venv/*
+
```
Comment (iangudger) [inline DB/.gitignore]: Is this an extra blank line?

File: DB/api/query_db.py [hunk 3]
```diff
@@ -287,13 +294,13 @@ def format_data(header, data, json=False):
...
+        yield '{"data": '
+    yield "[" + format_data_row(header)  # No comma before header
     for row in data:
-        yield ',\n' + format_data_row(row)
-    yield "\n]\n"
+        yield ',' + format_data_row(row)
+    yield "]"
     if json:
-        yield "}\n"
+        yield "}"
```
Comment (iangudger) [inline DB/api/query_db.py]: Why did you remove all of the newlines?

File: DB/api/server.py [hunk 4]
```diff
@@ -1,10 +1,12 @@
-#!/usr/bin/env python3
```
Comment (iangudger) [inline DB/api/server.py]: Why did you remove this? We need it to be able to execute the script directly.

File: DB/api/server.py [hunk 5]
```diff
@@ -1,10 +1,12 @@
...
+import sys
+import os
+# add Backend directory to python path so we can do
```
Comment (iangudger) [inline DB/api/server.py]: Add a newline before this comment.

File: DB/classification/__init__.py [hunk 6]
```diff
@@ -0,0 +1 @@
+
```
Comment (iangudger) [inline DB/classification/__init__.py]: Do we need this file?

File: DB/api/__init__.py [hunk 7]
```diff
@@ -1 +0,0 @@
-__all__ = ["Analysis_3", "query_db", "insert_db","detect_events", "handler", "server", "url_parser"]
```
Comment (iangudger) [inline DB/api/__init__.py]: Why this change?

File: .travis.yml [hunk 8]
```diff
@@ -12,14 +12,17 @@ env:
...
 
 addons:
   apt:
     packages:
     - libpq-dev
     - python3-numpy
+    - python3-scipy
     - python3-psycopg2
 
+
```
Comment (iangudger) [inline .travis.yml]: I don't think we need this extra blank line.

File: DB/api/query_db.py [hunk 9]
```diff
@@ -53,8 +56,17 @@ def query(parsed_url):
...
         events = parsed_url['events']
+    if 'total_energy' in parsed_url.keys():
+        total_energy = parsed_url['total_energy']
+
+    if classify:
+        if device is not None and start_time is not None:
+            model = RMF.RandomForestModel.get_model()
+            classification = model.classify(time=start_time, serial=device_id, panel=device)
+            return format_data(['data'], classification, json=True)
+        raise Exception("Received malformed URL data: missing start_time")
```
Comment (iangudger) [inline DB/api/query_db.py]: Why don't we care about the endtime any more?

File: DB/api/test_query_db.py [hunk 10]
```diff
@@ -84,6 +84,8 @@ def test_classify(self):
         self.retrieve_args[0] = 5
 
         self.query_options['classify'] = True
+        self.query_options['device'] = 'Panel1'
+        self.retrieve_args[7] = 'Panel1'
```
Comment (iangudger) [inline DB/api/test_query_db.py]: Better to put these in order. Ex self.retrieve_args[0], self.retrieve_args[6], self.retrieve_args[7]

File: DB/classification/__init__.py (hunk 6 above)
Comment (nsokolnikov) [inline DB/classification/__init__.py]: Yes.

File: DB/api/query_db.py (hunk 9 above)
Comment (nsokolnikov) [inline DB/api/query_db.py]: Because the end-time is determined by the window size and is generated internally inside the query execute method based on the window size of the model it's trying to classify with.

File: DB/api/test_query_db.py [hunk 11]
```diff
@@ -96,10 +98,12 @@ def test_classify(self):
```
Comment (iangudger) [inline DB/api/test_query_db.py]: The reason this test is failing is because it is still stubbing out the old analysis library. You need to change it from stubbing out Analysis_3.run to stubbing out RMF.RandomForestModel.get_model and model.classify.

File: DB/api/__init__.py (hunk 7 above)
Comment (ianlofs) [inline DB/api/__init__.py]: You dont actually need anything in the __init__ files, they are just there to structure the modules/packages

Comment (iangudger) [inline DB/api/__init__.py]: Is there a reason why we no longer want to initialize __all__?

File: DB/api/server.py (hunk 4 above)
Comment (ianlofs) [inline DB/api/server.py]: This was an accident

**Comment by nsokolnikov** [comment]
I want to merge this. I looked through and I would have to remove almost all of our tests/rewrite them entirely to stub out the database calls. Particularly test_query_db.py has db calls sprinkled throughout. 

Master build is failing so we might as well merge. The only failing parts are the tests, and they are failing for db access reasons. The rest of it works fine and the tests have been passing locally.

It needs to be merged so that frontend can access our classifiers for their development and we will focus on making travis pass in the meantime.

File: DB/api/query_db.py [hunk 12]
```diff
@@ -24,8 +27,8 @@ def query(parsed_url):
     device_id = parsed_url['device_id']
 
     header = ['time', 'I', 'W', 'V', 'T']
-    start_time = end_time = data_type = subset = limit = device = granularity = None
-    diff = json = reverse = classify = list_format = False
+    start_time = end_time = data_type = subset = limit = device = granularity =None
```
Comment (iangudger) [inline DB/api/query_db.py]: Please add the space back.

File: DB/api/server.py (hunk 4 above)
Comment (iangudger) [inline DB/api/server.py]: Can you add it back?

File: DB/api/test_query_db.py [hunk 13]
```diff
@@ -84,6 +84,8 @@ def test_classify(self):
         self.retrieve_args[0] = 5
```
Comment (iangudger) [inline DB/api/test_query_db.py]: Please update this test to reflect the fact that you replaced the Analysis_3 library.

File: DB/classification/models.py [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+    @Memoized
+    def get_model():
+        try:
+            con = psycopg2.connect(database=DATABASE, user=USER)
+        except psycopg2.Error as e:
+            raise ConnectionError("Database connection on model insert", e)
+
+        try:
+            cursor = con.cursor()
+            query = "SELECT * FROM classifier_model ORDER BY created_at DESC LIMIT 1;"
```
Comment (iangudger) [inline DB/classification/models.py]: You shouldn't use SELECT * in code. If new columns are added, your code breaks.

**Comment by aadabi** [comment]
Mr Nikita mentioned Ian approved to ignore Circle Ci stuff failure for now.

File: .travis.yml (hunk 1 above)
Comment (nsokolnikov) [inline .travis.yml]: Yes, it imports the python modules in the backend.
